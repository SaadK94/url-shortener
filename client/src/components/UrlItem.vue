<template>
  <div class="container-outer">
    <div class="container-inner">
      <a class="full-url" :href="fullUrl" target="_blank" data-jest-full-url>{{
        fullUrl
      }}</a>
      <h4 class="short-url" data-jest-short-url>{{ shortUrl }}</h4>
    </div>
    <div class="buttons">
      <button class="copy" @click="copyUrl" data-jest-copy>Copy</button>
      <button class="delete" @click="removeUrl" data-jest-delete>Delete</button>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { mapActions } from "vuex";

export default defineComponent({
  props: {
    id: {
      type: String,
      required: true,
    },
    fullUrl: {
      type: String,
      required: true,
    },
    shortUrl: {
      type: String,
      required: true,
    },
  },
  methods: {
    ...mapActions(["deleteUrl", "showSnackbar"]),
    async removeUrl() {
      await this.deleteUrl(this.id);
      await this.showSnackbar("URL deleted");
    },
    async copyUrl() {
      await navigator.clipboard.writeText(this.shortUrl);
      await this.showSnackbar("URL copied to clipboard");
    },
  },
});
</script>

<style scoped>
.container-outer {
  display: flex;
  align-items: center;
  justify-content: center;
}

.container-inner {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  background-color: #f4f5f5;
  border-radius: 8px;
  overflow: auto;
  flex-grow: 1;
}

.buttons {
  display: flex;
  flex-direction: column;
}

.full-url {
  text-decoration: none;
  color: #5bb0b8;
  font-weight: bold;
  margin: auto;
  padding: 16px;
}

.full-url:hover {
  color: #69d1db;
}

.short-url {
  margin: auto;
  color: #27515f;
  padding: 6px;
}

button {
  padding: 10px;
  border-radius: 8px;
  cursor: pointer;
  color: white;
  font-size: 17px;
  text-align: center;
  background-color: #27515f;
  width: 175px;
  margin: 0 0 4px 16px;
}

button.copy {
  background-color: #5bb0b8;
}

button.delete {
  background-color: #ba3030;
}

button.delete:hover {
  background-color: #e33d3d;
}

button.copy:hover {
  background-color: #a9d5d9;
}

@media (max-width: 768px) {
  .container-outer,
  .container-inner {
    flex-direction: column;
    align-items: stretch;
  }

  .full-url {
    margin: 12px auto;
    padding: 0 6px;
  }
  .short-url {
    margin: auto;
    padding: 10px 0;
  }

  button {
    width: 100%;
    margin: 12px 0 0 0;
  }
}
</style>
