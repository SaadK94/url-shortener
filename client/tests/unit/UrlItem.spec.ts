import { shallowMount, VueWrapper } from "@vue/test-utils";
import UrlItem from "@/components/UrlItem.vue";

describe("UrlItem.vue", () => {
  let wrapper: VueWrapper<any>;

  const props = {
    id: "123",
    fullUrl: "https://www.google.co.uk",
    shortUrl: "https://pbid.io/abcd1234",
  };

  beforeEach(() => {
    wrapper = shallowMount(UrlItem, { props });
  });

  it("renders the component correctly with the given props", () => {
    expect(wrapper.find("[data-jest-full-url]").text()).toEqual(
      "https://www.google.co.uk"
    );
    expect(wrapper.find("[data-jest-short-url]").text()).toEqual(
      "https://pbid.io/abcd1234"
    );
  });

  it("calls the correct method on each button click", async () => {
    wrapper.vm.copyUrl = jest.fn();
    wrapper.vm.removeUrl = jest.fn();
    await wrapper.find("[data-jest-copy]").trigger("click");
    expect(wrapper.vm.copyUrl).toHaveBeenCalled();
    await wrapper.find("[data-jest-delete]").trigger("click");
    expect(wrapper.vm.removeUrl).toHaveBeenCalled();
  });

  it("matches the snapshot", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
