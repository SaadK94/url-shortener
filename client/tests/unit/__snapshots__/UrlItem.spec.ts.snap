// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlItem.vue matches the snapshot 1`] = `
<div
  class="container-outer"
>
  <div
    class="container-inner"
  >
    <a
      class="full-url"
      data-jest-full-url=""
      href="https://www.google.co.uk"
      target="_blank"
    >
      https://www.google.co.uk
    </a>
    <h4
      class="short-url"
      data-jest-short-url=""
    >
      https://pbid.io/abcd1234
    </h4>
  </div>
  <div
    class="buttons"
  >
    <button
      class="copy"
      data-jest-copy=""
    >
      Copy
    </button>
    <button
      class="delete"
      data-jest-delete=""
    >
      Delete
    </button>
  </div>
</div>
`;
